### PR TITLE
Added option to specify you want the keyboard to be dismissed when you scroll.

### PR DESCRIPTION
**Pull request by gaaclarke**
## Description

Created an option when creating ListViews and ScrollViews that can control how a ScrollView can automatically dismiss the keyboard (a la [UIScrollViewKeyboardDismissMode ](https://developer.apple.com/documentation/uikit/uiscrollviewkeyboarddismissmode?language=objc)).

## Related Issues

https://github.com/flutter/flutter/issues/21814

## Tests

I added the following tests:

*Replace this with a list of the tests that you added as part of this PR. A change in behaviour with no test covering it
will likely get reverted accidentally sooner or later. PRs must include tests for all changed/updated/fixed behaviors. See [Test Coverage].*

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: *Replace with a link to your migration guide*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by gaaclarke** [comment]
@gspencergoog Can you check out what I'm doing here to make sure there isn't a different way we'd want to do this?  Basically we want to make it easy for people to specify how a keyboard should be dismissed based on a scrollview's interaction.  It's the equivalent to this: https://developer.apple.com/documentation/uikit/uiscrollviewkeyboarddismissmode/uiscrollviewkeyboarddismissmodeondrag?language=objc

File: packages/flutter/lib/src/widgets/scroll_view.dart
Comment (gspencergoog) [inline packages/flutter/lib/src/widgets/scroll_view.dart]: You don't want to recreate this `FocusNode` every build, because you need to dispose of it, which (sadly) means you'd need to make this a stateful widget.

But... What's your goal here? Do you just want to know if a context doesn't have focus?  If so, you could use something like this, and keep it stateless.

```dart
  return Focus(
    child: Builder(builder: (BuildContext context) {
      return NotificationListener<ScrollUpdateNotification>(
        child: scrollableResult,
        onNotification: (ScrollUpdateNotification notification) {
          final FocusNode node = Focus.of(context);
          if (notification.dragDetails != null && !node.hasFocus) {
            node.requestFocus();
          }
          return true;
        },
      );
    }),
  );
```

Comment (gaaclarke) [reply packages/flutter/lib/src/widgets/scroll_view.dart]: For the focus node, I'm using that to dismiss the keyboard when scrolling happens.  I don't want to spam the Focus every time a scroll event happens so I keep a reference to the dismissNode to make sure that I only requestFocus once.  It's fine if the FocusNode gets disposed, it's just a way to filter out duplicate events, if it never worked everything would still look the same.

**Comment by gaaclarke** [comment]
@gspencergoog I just wanted to get a gut check that we are fine adding a parameter to ScrollView for the keyboard dismiss behavior.  In the future I'd like to add a third option which is "interactive" to duplicate iOS's native behavior.

File: packages/flutter/lib/src/widgets/scroll_view.dart
Comment (dnfield) [inline packages/flutter/lib/src/widgets/scroll_view.dart]: We probably shouldn't return true here -eating scroll notifications this low in the tree is likely to break people further up who also want to know about scrolling.

File: packages/flutter/lib/src/widgets/scroll_view.dart
Comment (dnfield) [reply packages/flutter/lib/src/widgets/scroll_view.dart]: Where are you disposing it?

Comment (gaaclarke) [reply packages/flutter/lib/src/widgets/scroll_view.dart]: I didn't realize we were talking about literally calling `.dispose()`.  I'll rework it.

Comment (gaaclarke) [reply packages/flutter/lib/src/widgets/scroll_view.dart]: Started using unfocus instead so no FocusNode instance is needed.

File: packages/flutter/lib/src/widgets/scroll_view.dart
Comment (gaaclarke) [inline packages/flutter/lib/src/widgets/scroll_view.dart]: Apple calls this behavior "none", I chose "manual".

File: packages/flutter/lib/src/widgets/scroll_view.dart
Comment (gspencergoog) [inline packages/flutter/lib/src/widgets/scroll_view.dart]: Why do you need this to write this differently from `focusScope.hasFocus`?  And if that's the case, then why wouldn't this work?:

```dart
if (notification.dragDetails != null) {
  focusScope.unfocus();
}
```

Basically, what's the reason for the extra conditions?

Comment (gaaclarke) [reply packages/flutter/lib/src/widgets/scroll_view.dart]: `notifcation.dragDetails != null` is to make sure the scrolling is the result of a drag, these events get fired too when the scrollview is scrolled as the result of a keyboard showing.

`focusScope.focusedChild != null` is because I'm going to read from the focusedChild and I don't want to read from a null pointer.

`focusScope.focusedChild.hasFocus` makes sure I don't spam `focusScope.unfocus()` and that it only gets called once for all the scrolling events that get sent.

Comment (gaaclarke) [reply packages/flutter/lib/src/widgets/scroll_view.dart]: Could I just used `focusScole.hasFocus` and avoid talking to the focusedChild?

Comment (gaaclarke) [reply packages/flutter/lib/src/widgets/scroll_view.dart]: I switched it over to `focusScope.hasFocus`.  It wasn't like this because i found the other way to do it first and it worked.

**Comment by gaaclarke** [comment]
@gspencergoog Oops, i just realized that changing from a Draft to a PR but it didn't remove the "WIP" label.  I'm ready to merge this when approved.

File: packages/flutter/lib/src/widgets/scroll_view.dart
Comment (gspencergoog) [inline packages/flutter/lib/src/widgets/scroll_view.dart]: Is there a reason you want to see if the scope is focused, rather than the nearest focus node? It's possible that this `ScrollView` isn't the focused child of the scope: if there's another child of the scope (i.e. a sibling of the `ScrollView`), and it's focused, then `hasFocus` would return true, but children of this `ScrollView` wouldn't have focus.

If you expect the `ScrollView` to be enclosed in a `Focus` widget, you could use `Focus.of(context).hasFocus`. If you don't expect that (which is probably the case), you can introduce a focus node here (probably using a `Focus` widget with `canRequestFocus` set to false) and test to see if that has focus, which would mean that one of the children of the `ScrollView` has focus. You can keep the widget stateless by using a `Builder` to let you look up the node that the `Focus` introduces, as in my previous comment.

Comment (gaaclarke) [reply packages/flutter/lib/src/widgets/scroll_view.dart]: Here is what I would like to write:
```
if (Keyboard.isVisible) {
  Keyboard.hide();
}
```

If the scrollview that has a sibling that is a TextField that is focused, we want this scrolling action to dismiss his keyboard too.  (I think that is the case you are mentioning).

Comment (gspencergoog) [reply packages/flutter/lib/src/widgets/scroll_view.dart]: OK, in that case, then looking at the scope is appropriate.

**Comment by gaaclarke** [comment]
@gspencergoog can I get an LGTM?

**Review by gspencergoog — APPROVED** [review]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

**Pull request merged**
